modules/asset-manager-impl/src/main/java/org/opencastproject/assetmanager/impl/AbstractAssetManagerWithTieredStorage.java

```diff
@@ -0,0 +1,412 @@
+  private boolean isRemoteAssetStoreId(String storeId) {
+    return !isLocalAssetStoreId(storeId);
+  }
+
+  @Override
+  public Opt<Asset> getAsset(Version version, final String mpId, final String mpeId) {
+    // try to fetch the asset
+    for (final AssetDtos.Medium asset : getDb().getAsset(RuntimeTypes.convert(version), mpId, mpeId)) {
+      for (final String storageId : getSnapshotStorageLocation(version, mpId)) {
+        for (final AssetStore store : getAssetStore(storageId)) {
+          for (final InputStream assetStream : store.get(StoragePath.mk(asset.getOrganizationId(), mpId, version, mpeId))) {
+            final Asset a = new AssetImpl(
+                    AssetId.mk(version, mpId, mpeId),
+                    assetStream,
+                    asset.getAssetDto().getMimeType(),
+                    asset.getAssetDto().getSize(),
+                    asset.getStorageId(),
+                    asset.getAvailability());
+            return Opt.some(a);
+          }
+        }
+      }
+    }
+    return Opt.none();
+  }
+
+  /** Move the assets for a snapshot to the target store */
+  private void copyAssetsToStore(Snapshot snap, AssetStore store) throws Exception {
+    final String mpId = snap.getMediaPackage().getIdentifier().toString();
+    final String orgId = snap.getOrganizationId();
+    final Version version = snap.getVersion();
+    final String prettyMpId = mpId + "@v" + version;
+    logger.debug(format("Moving assets for snapshot %s to store %s", prettyMpId, store.getStoreType()));
+    for (final MediaPackageElement e : snap.getMediaPackage().getElements()) {
+      if (!MOVABLE_TYPES.contains(e.getElementType())) {
+        logger.debug(format("Skipping %s because type is %s", e.getIdentifier(), e.getElementType()));
+        continue;
+      }
+      logger.debug(format("Moving %s to store %s", e.getIdentifier(), store.getStoreType()));
+      final StoragePath storagePath = StoragePath.mk(orgId, mpId, version, e.getIdentifier());
+      final Opt<StoragePath> existingAssetOpt = findAssetInVersionsAndStores(e.getChecksum().toString(), store.getStoreType());
+      if (existingAssetOpt.isSome()) {
+        final StoragePath existingAsset = existingAssetOpt.get();
+        logger.debug("Content of asset {} with checksum {} already exists in {}",
+                existingAsset.getMediaPackageElementId(), e.getChecksum(), store.getStoreType());
+        if (!store.copy(existingAsset, storagePath)) {
+          throw new AssetManagerException(
+                  format("An asset with checksum %s has already been archived but trying to copy or link asset %s to it failed",
+                          e.getChecksum(), existingAsset));
+        }
+      } else {
+        final Opt<Long> size = e.getSize() > 0 ? Opt.some(e.getSize()) : Opt.<Long>none();
+        store.put(storagePath, Source.mk(e.getURI(), size, Opt.nul(e.getMimeType())));
+      }
```

Comment:
The asset storeId in the database is only updated after all assets in the media package are moved. This will also cause assets to take more space in the remote store than they should. For instance, if a media package has 2 references to the same media, such as presenter/source and presenter/work, 2 files will be put in the remote store instead of having a logic hard-link between them.

Reply:
Line 369 (getDb().setAssetStorageLocation) is at the bottom of the for loop, which is iterating over the list of media package elements.  Barring two versions of a given snapshot being moved at the same time, the asset's storage ID should already be set correctly if it was just moved, wouldn't it?

Reply:
Well, it's this piece of code that decides between a 'copy' or a 'move' so it is this code's responsibility to tell the store if that element is a new one or not. The store's particular implementation may do linking or not.
Anyway, I think my PR addressed that too?

Reply:
I think so.  We'll see in further testing :D